libs/wingc/src/type_check.rs

```diff
@@ -1691,10 +1685,13 @@ impl Types {
 		self.json_literal_casts.get(&expr_id)
 	}
 
-	pub fn reset_expr_types(&mut self) {
-		for elem in self.type_for_expr.iter_mut() {
-			*elem = None;
-		}
+	pub fn reset_per_compile_data(&mut self) {
+		self.type_for_expr.clear();
+		self.scope_envs.clear();
+		self.inferences.clear();
+		self.json_literal_casts.clear();
+		self.type_expressions.clear();
+		self.source_file_envs.clear();
 	}
```

Comment:
I'm wondering if the approach in this commit would be more efficient, since maybe we'd be saving some time spent reloading type information about jsii libraries from `TypeSystem` into `Types`. For example, [here](https://github.com/winglang/wing/blob/7f087c73f1f79a6af94324f3b91ed17f14b1e97b/libs/wingc/src/lib.rs#L229-L241) inside the top-level `type_check()` function called by both `compile()` and `partial_compile()`, we're resetting reloading information about the SDK and built-in functions etc., and in theory those calls could be skipped or made into no-ops during re-compiles.

Anyways it could be worth checking out if there's any performance difference between these approaches, but otherwise either looks okay to me (maybe throwing away `Types` each time will simply be more error-proof in the long run)

Reply:
I think we could design around this, but it's tricky since we haven't been considering it strongly so far.
I'm pretty sure this approach causes a memory leak in the LSP for most type info because we'd never actually free anything that we don't have a good way to reuse.

Luckily the most expensive part of all of this (JSII importing) is saved already so I think it's okay to go with the safer/slower option for `Types` for now.